sound/soc/sof/pcm.c

```diff
@@ -249,7 +250,7 @@ static int sof_pcm_hw_params(struct snd_soc_component *component,
 
 	/* if this is a repeated hw_params without hw_free, skip setting up widgets */
 	if (!spcm->stream[substream->stream].list) {
-		ret = sof_pcm_setup_connected_widgets(sdev, rtd, spcm, substream->stream);
+		ret = sof_pcm_setup_connected_widgets(sdev, rtd, spcm, substream->stream, &pcm);
```

Comment:
completely lost here. the pcm structure is for the front-end, why would we need to pass this to widgets? what does this have to do with the SSP configuration at all?

Reply:
@plbossart That's a good comment. We should not assume FE params are the same, and rather we should use BE params for matching. Let's look at that first.

Then there's the bigger question of using a kcontrol/enum both @plbossart and @lgirdwood you raised. Now this is something I have pushed and we should have discussed this with more attention sooner. The way I view it:

- HW supports a finite set of sampling rates (as defined in topology)
- user-space will indicate which sampling rate it wants to use via ALSA PCM interface

Plusses and minuses for this design:
 + This has the benefit that the interface has nothing specific to SOF or Intel HDA, but is a generic interface all applications already use, and matches how Bluetooth PCM is exposed by other hw vendors. 
 + This is aligned with how ALSA PCM interface works -- hardware can be configured via hwparams and there is mechanism to report to user-space what was chosen.
 + A BT profile change (moving from music listening to a voice call), is simply ALSA PCM reconfiguration, done wihtin the PCM FE state machine. I.e. application needs to close the PCM and reopen with params of the new profile.
 -  This design will not work if a SRC is put to the pipeline and front-end sampling rate is not related to SSP sampling rate. This is a choice we make when defining the DSP topology, and I'd argue this is still reasonable. If we'd implement a SRC pipeline in DSP, we need some new Intel specific interface, making the overall solution less generic.

Plusses/minuses for changing the design to control SSP config via enum kcontrol:

+ Allows adding SRC to pipeline.
- User-space has to use hw-specific interface to switch the SSP config.
- When BT profile changes, user-space needs to execute a coordinated sequence of PCM and kcontrol actions, which again is hw specific. 
+ We should be able to describe the actions with UCM, enabling to hide the hw specifics.